### DISCUSSION TITLE
CI

### DISCUSSION THREAD

**Discussion by felix-seifert** [csgoh/roadmapper#16] [Ideas]
Before each PR is merged into the main branch, the merging entity, quite likely the maintainer, has to either run the test suite manually or trust that the changes do not break the main branch. 

The idea is to let the tests of roadmapper run automatically before a PR can be merged. Additionally, these checks could be run on different platforms (e.g. Mac, Ubuntu and Windows).

@csgoh Do you want to explain the testing cycle that should be performed? I can then implement this with GitHub actions, and we make the passing of the CI tests a branch protection rule.

**Comment by csgoh** [comment]
So far, I am the only one maintaining roadmapper; the testing process is being done manually. One of the difficulties is that the output is always in PNG format; unless I compare the actual result to the expected result in pixels, the automated test cannot go very far.

I believe that automating PR testing across multiple platforms is beneficial.

As I cross the bridge (if there are more than 1 maintainer), I guess I will have to look at the automated CI/CD workflow.

Do you wish to assist and contribute to this repository? 😉

**Reply by felix-seifert** [reply]
The comparison of the generated PNGs can be achieved with Selenium. Doing this kind of E2E tests is very helpful in ensuring that the product delivers what it promises. Nevertheless, the testing should be more granular with unit tests testing the function isolated from other functionalities. Even though this seems unnecessary if you already have E2E tests, it is easier to debug coding mistakes down the line and you can cover more special cases. We might then have to have a look at providing a proper dir structure.

To answer your last question: Yes, I like to contribute to this repository. I might not be too much of a feature developer but more of an infrastructure dev. My idea is to work on the CI workflow and I already have some plans for a declarative language to build roadmaps without coding knowledge.

**Reply by csgoh** [reply]
Invited you as collaborator. Any help to improve the repo is much appreciated.
